chia/types/eligible_coin_spends.py

```diff
@@ -97,12 +219,145 @@ def get_deduplication_info(
+            # We've added a ff spend with this puzzle hash before, so build on that
+            # NOTE: As it's not possible to submit a transaction to the mempool that
+            # spends the output of another transaction already in the mempool,
+            # we don't need to check if we're the most recent version because
+            # at this point we cannot be, so we must fast forward
+            new_coin_spend, patched_additions = perform_the_fast_forward(
+                unspent_lineage_info=unspent_lineage_info,
+                spend_data=spend_data,
+                fast_forward_spends=self.fast_forward_spends,
+            )
+            # Mark this coin for a coin spend data update
+            replaced_coin_ids.append(coin_id)
+            ff_bundle_coin_spends[new_coin_spend.coin.name()] = BundleCoinSpend(
+                coin_spend=new_coin_spend,
+                eligible_for_dedup=spend_data.eligible_for_dedup,
+                eligible_for_fast_forward=spend_data.eligible_for_fast_forward,
+                additions=patched_additions,
+                cost=spend_data.cost,
+            )
+            # Update the list of coins spends that make the new fast forward bundle
+            new_coin_spends.append(new_coin_spend)
+        if len(ff_bundle_coin_spends) == 0:
+            # This item doesn't have any fast forward coins, nothing to do here
+            return
+        # Update the mempool item after validating the new spend bundle
+        new_sb = SpendBundle(
+            coin_spends=new_coin_spends, aggregated_signature=mempool_item.spend_bundle.aggregated_signature
+        )
+        # We need to run the new spend bundle to make sure it remains valid
+        generator = simple_solution_generator(new_sb)
+        assert mempool_item.npc_result.conds is not None
+        new_npc_result = get_name_puzzle_conditions(
+            generator=generator,
+            max_cost=mempool_item.npc_result.conds.cost,
+            mempool_mode=True,
+            height=height,
+            constants=constants,
+        )
+        if new_npc_result.error is not None:
+            raise ValueError("Mempool item became invalid after singleton fast forward.")
+        # Update bundle_coin_spends using the collected data
+        for coin_id in replaced_coin_ids:
+            mempool_item.bundle_coin_spends.pop(coin_id, None)
+        mempool_item.bundle_coin_spends.update(ff_bundle_coin_spends)
+        # Update the mempool item with the new spend bundle related information
+        # NOTE: From this point on, in `create_bundle_from_mempool_items`, we rely
+        # on `bundle_coin_spends` and we don't use this updated spend bundle
+        # information, as we'll only need `aggregated_signature` which doesn't
+        # change. Still, it's good form to update the spend bundle with the
+        # new coin spends
+        mempool_item = dataclasses.replace(mempool_item, spend_bundle=new_sb, npc_result=new_npc_result)
```

Comment:
are you sure this works? my understanding of python is that `mempool_item` is a local variable that points to the object that was passed in. on this line, all you do is changing what the *local* variable `mempool_item` points to, you don't alter the object passed in.

Reply:
The code now explicitly mutates the internal mempool item.